protobuf/v1/nexus.proto

```diff
@@ -0,0 +1,155 @@
+  rpc RemoveChildNexus (RemoveChildNexusRequest) returns (google.protobuf.Empty) {}
+  rpc FaultNexusChild (FaultNexusChildRequest) returns (google.protobuf.Empty) {}
+
+  // This method is called by control plane to construct a block device
+  // (/dev/...) that will be used to connect the nexus to the OS.
+  rpc PublishNexus (PublishNexusRequest) returns (PublishNexusResponse) {}
+  rpc UnpublishNexus (UnpublishNexusRequest) returns (google.protobuf.Empty) {}
+
+  // NVMe ANA state
+  rpc GetNvmeAnaState (GetNvmeAnaStateRequest) returns (GetNvmeAnaStateResponse) {}
+  rpc SetNvmeAnaState (SetNvmeAnaStateRequest) returns (google.protobuf.Empty) {}
+
+}
+
+// Create nexus arguments.
+message CreateNexusRequest {
+  string name = 1; // name of the nexus
+  string uuid = 2; // UUID of the nexus
+  uint64 size = 3; // size of the device in bytes
+  uint32 minCntlId = 4;  // minimum NVMe controller ID
+  uint32 maxCntlId = 5;  // maximum NVMe controller ID
+  uint64 resvKey = 6;    // NVMe reservation key for children
+  uint64 preemptKey = 7; // NVMe preempt key for children
+  repeated string children = 8; // uris to the targets we connect to
+}
+
+// State of the nexus child.
+enum ChildState {
+  CHILD_UNKNOWN = 0;
+  CHILD_ONLINE = 1;   // healthy and contains the latest bits
+  CHILD_DEGRADED = 2; // rebuild is in progress (or other recoverable error)
+  CHILD_FAULTED = 3;  // unrecoverable error (control plane must act)
+}
+
+// represents a child device part of a nexus
+message Child {
+  string uri = 1;   // uri of the child device
+  ChildState state = 2; // state of the child
+  int32 rebuild_progress = 3;
+}
+
+// State of the nexus (terminology inspired by ZFS).
+enum NexusState {
+  NEXUS_UNKNOWN = 0;
+  NEXUS_ONLINE = 1;    // healthy and working
+  NEXUS_DEGRADED = 2;  // not healthy but is able to serve IO (i.e. rebuild is in progress)
+  NEXUS_FAULTED = 3;   // broken and unable to serve IO
+}
+
+enum NvmeAnaState {
+  NVME_ANA_INVALID_STATE = 0;           // invalid, do not use
+  NVME_ANA_OPTIMIZED_STATE = 0x1;
+  NVME_ANA_NON_OPTIMIZED_STATE = 0x2;
+  NVME_ANA_INACCESSIBLE_STATE = 0x3;
+  NVME_ANA_PERSISTENT_LOSS_STATE = 0x4; // not yet supported
+  NVME_ANA_CHANGE_STATE = 0xF;          // not yet supported
+}
+
+// represents a nexus device
+message Nexus {
+  string name = 1;             // name of the nexus
+  string uuid = 2;             // UUID of the bdev
+  uint64 size = 3;             // size of the volume in bytes
+  NexusState state = 4;        // current state of the nexus
+  repeated Child children = 5; // array of children
+  // URI of the device for the volume (missing if not published).
+  // Missing property and empty string are treated the same.
+  string device_uri = 6;
+  uint32 rebuilds = 7;         // total number of rebuild tasks
+  NvmeAnaState ana_state = 8;  // Nexus ANA state.
+}
+
+message CreateNexusResponse {
+  Nexus nexus = 1;
+}
+
+message AddChildNexusResponse {
+  Nexus nexus = 1;
```

Comment:
```suggestion
  Child nexus = 1;
```

Reply:
IMHO it should be nexus - in the control plane everytime we have a nexus child operation we have to fetch the nexus right after the call. If the call returns a Nexus then we already have everything we need as the nexus contains the child.